org.eclipse.dawnsci.analysis.dataset.test/src/org/eclipse/dawnsci/analysis/dataset/FunctionalUtilsTest.java

```diff
@@ -0,0 +1,210 @@
+package org.eclipse.dawnsci.analysis.dataset;
+
+import static org.junit.Assert.assertArrayEquals;
+import static org.junit.Assert.assertEquals;
+
+import java.util.stream.IntStream;
+
+import org.eclipse.january.dataset.ByteDataset;
+import org.eclipse.january.dataset.Dataset;
+import org.eclipse.january.dataset.DatasetFactory;
+import org.eclipse.january.dataset.DoubleDataset;
+import org.eclipse.january.dataset.FloatDataset;
+import org.eclipse.january.dataset.IndexIterator;
+import org.eclipse.january.dataset.IntegerDataset;
+import org.eclipse.january.dataset.LongDataset;
+import org.eclipse.january.dataset.Random;
+import org.eclipse.january.dataset.ShortDataset;
+import org.eclipse.january.dataset.Slice;
+import org.junit.Test;
+
+public class FunctionalUtilsTest {
+
+	@Test
+	public void testDatasetIndexFunction() {
+		Dataset a = DatasetFactory.createRange(144).reshape(12, -1);
+		generateAndCheckFunction(a);
+		generateAndCheckFunction(a.getTransposedView());
+
+		Dataset v = a.getSliceView(new Slice(3, 10), new Slice(null, null, -2));
+		generateAndCheckFunction(v);
+		generateAndCheckFunction(v.getTransposedView());
+
+		Dataset z = a.getSliceView(new Slice(3, 3), new Slice(1, 1));
+		assertEquals(0, z.getSize());
+		generateAndCheckFunction(z);
+		generateAndCheckFunction(z.getTransposedView());
+	}
+
+	private int[] generateIndexes(Dataset v) {
+		int[] eIndex = new int[v.getSize()];
+		IndexIterator it = v.getIterator();
+		for (int i = 0; it.hasNext(); i++) {
```

Comment:
![MAJOR](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/severity-major.png 'Severity: MAJOR') This loop's stop condition tests "it" but the incrementer updates "i". [![rule](https://raw.githubusercontent.com/SonarCommunity/sonar-github/master/images/rule.png)](https://sonarqube.com/coding_rules#rule_key=squid%3AS1994)